# -62 is N2, wildtype
# -82 is r910, smg-1 mutant

SAMPLES=[
	"SRR5811862.subset"
]

rule all:
	input:
		expand("HTS_testing/{sample}.streamed.json", sample=SAMPLES),
		expand("HTS_testing/{sample}.streamed_SE.fastq.gz", sample=SAMPLES)
		
'''
# gets basic statistics about read data
rule basic_stats:
	input:
		"data/subset/{sample}.fastq.gz"
	output:
		multiext("HTS_testing/{sample}", ".stats.json", ".stats_SE.fastq.gz")
	shell:
		"hts_Stats -U {input} -L HTS_testing/{wildcards.sample}.stats.json -f "
		"HTS_testing/{wildcards.sample}.stats"
		
# count ribosomal rna
# rrna ref sequences from NCBI
rule count_rrna:
	input:
		"HTS_testing/{sample}.stats.json"
		fastq = "data/subset/SRR5811862.subset.fastq.gz",
		ref = "references/celegans_rrna.fasta"
	output:
		multiext("HTS_testing/SRR5811862.subset", ".rrna.json", ".rrna_SE.fastq.gz")
	shell:
		"hts_SeqScreener -A {input}
		

'''
# pipe hts_Stats to hts_SeqScreener
# runs basic stats and counts rrna (rrna ref seq from NCBI)
# phix reads screened out by default
rule stream_hts:
	input:
		mrna = "data/subset/{sample}.fastq.gz",
		rrna = "references/celegans_rrna.fasta"
	output:
		multiext("HTS_testing/{sample}", ".streamed.json", \
		".streamed_SE.fastq.gz")
	shell:
		"hts_Stats -U {input.mrna} -L "
		"HTS_testing/{wildcards.sample}.streamed.json | hts_SeqScreener -A "
		"HTS_testing/{wildcards.sample}.streamed.json -r -s {input.rrna} -f "
		"HTS_testing/{wildcards.sample}.streamed"
		
	
	
	
	
	
	
	
	
	
	
	
	
	
	
	
	
	
	
