SAMPLES=[
	"SRR5811862.subset",
	"SRR5811882.subset"
]

EXT1S=["1.ht2", "2.ht2", "3.ht2", "4.ht2", "5.ht2", "6.ht2", "7.ht2", "8.ht2"]

EXT2S=["sam", "junctions"]

GENOME=["1pct_elegans"]

REFrRNA="references/celegans_rrna.fasta"
SAMPLEmRNA=expand("data/subset/{sample}.fastq.gz", sample=SAMPLES)
REFGEN="references/1pct_elegans.fa"
STAR_genomeSAindexNbases=8

rule all:
	input:
		expand("HTS_output/{sample}.htsStats.json", sample=SAMPLES),
		expand("HTS_output/{sample}.htstream_SE.fastq.gz", sample=SAMPLES),
		directory("star_index/")

rule stream_hts:
	output:
		multiext("HTS_output/{sample}", ".htsStats.json", \
		".htstream_SE.fastq.gz")
	params:
		rrna=REFrRNA,
		mrna=SAMPLEmRNA
	shell:
		"hts_Stats -L HTS_output/{wildcards.sample}.htsStats.json "
		"-N 'initial stats' -U {params.mrna} | "
		"hts_SeqScreener -A HTS_output/{wildcards.sample}.htsStats.json "
		"-N 'screen phix' | "
		"hts_SeqScreener -A HTS_output/{wildcards.sample}.htsStats.json "
		"-N 'count the number of rRNA reads' -r -s {params.rrna} | "
		"hts_PolyATTrim -A HTS_output/{wildcards.sample}.htsStats.json "
		"-N 'trim adapters' | "
		"hts_NTrimmer -A HTS_output/{wildcards.sample}.htsStats.json "
		"-N 'remove any remaining 'N' characters' | "
		"hts_QWindowTrim -A HTS_output/{wildcards.sample}.htsStats.json "
		"-N 'quality trim the ends of reads' | "
		"hts_LengthFilter -A HTS_output/{wildcards.sample}.htsStats.json "
		"-N 'remove reads < 50 bp' -n -m 50 | "
		"hts_Stats -A HTS_output/{wildcards.sample}.htsStats.json "
		"-f HTS_output/{wildcards.sample}.htstream > HTS_output/out.tab"

rule build_star_index:
	output:
		directory("star_index/")
	params:
		ref=REFGEN,
		str_len=STAR_genomeSAindexNbases
	shell:
		"STAR --runMode genomeGenerate --genomeSAindexNbases {params.str_len} "
		"--genomeDir star_index/ --genomeFastaFiles {params.ref} --sjdbGTFfile "
		"references/1pct_elegans.gffread.gtf"
		



'''
		"star_index/chrLength.txt",
		"star_index/chrNameLength.txt",
		"star_index/chrName.txt",
		"star_index/chrStart.txt",
		"star_index/exonGeTrInfo.tab",
		"star_index/exonInfo.tab",
		"star_index/geneInfo.tab",
		"star_index/Genome",
		"star_index/genomeParameters.txt",
		"star_index/Log.out",
		"star_index/SA",
		"star_index/SAindex",
		"star_index/sjdbInfo.txt",
		"star_index/sjdbList.fromGTF.out.tab",
		"star_index/sjdbList.out.tab",
		"star_index/transcriptInfo.tab"
'''













